hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-jobclient/src/test/java/org/apache/hadoop/mapred/pipes/TestPipeApplication.java

```diff
@@ -455,6 +460,96 @@ public void testPipesPartitioner() {
     assertEquals(3, partitioner.getPartition(iw, new Text("test"), 2));
   }
 
+  @Test
+  public void testSocketCleaner() throws Exception {
+    ServerSocket serverSocket = setupServerSocket();
+    SocketCleaner cleaner = setupCleaner(serverSocket);
+    // mock ping thread, connect to server socket per second.
+    int expectedClosedCount = 5;
+    for (int i = 0; i < expectedClosedCount; i++) {
+      try {
+        Thread.sleep(1000);
+        Socket clientSocket = new Socket(serverSocket.getInetAddress(),
+                                         serverSocket.getLocalPort());
+        clientSocket.close();
+      } catch (Exception exception) {
+        // ignored...
+        exception.printStackTrace();
+      }
+    }
+    waitUntilExhausted(
+        () -> expectedClosedCount == cleaner.getCloseSocketCount(), 5000);
```

Comment:
GenericTestUtils can be used instead.
```suggestion
    GenericTestUtils.waitFor(
        () -> expectedClosedCount == cleaner.getCloseSocketCount(), 100, 5000);
```

Reply:
Thanks for advice!